packages/material-ui/test/typescript/styles.spec.tsx

```diff
@@ -354,3 +355,50 @@ withStyles(theme =>
     text: theme.typography.body2,
   });
 }
+
+{
+  // can't provide own `classes` type
+  interface Props {
+    classes: number;
+  }
+
+  class Component extends React.Component<Props & WithStyles<typeof styles>> {}
+  // $ExpectError
+  const StyledComponent = withStyles(styles)(Component);
```

Comment:
@pelotom This is a limitation of the `never` type. I cannot assign T to never but a function that takes T to a function that takes never. [typescript playground](https://bit.ly/2NhVNs0)

If I type this function explicitly as a `React.SFC` I do get errors but the message is confusing:
`Property 'children' is missing in type 'ValidationMap<Props>'`.

I'm going to verify if this was working before. If we can't resolve this we have to make a decision whether we want to support `defaultProps` or the consistency check for stateless components.

Reply:
It did indeed work. 

I would prefer `defaultProps` support over consistency checks for implicit SFCs. We could always add a section to the typescript guide.

Reply:
Not with the `never` approach. That's either an issue with ts (although I couldn't really find reports on this) or an actual limitation. I tried using the previous approach with ConsistentWith but either ts inferred `{}` as the prop type or I got errors when passing a styled component to `withTheme`

I probably revisit the react-redux typings since I used the `Matching` from there and see if I can connect inconsistent props with their typings too. That way we get more focus on the issue and maybe get some more help.

But as is the issue with #12697 this problem at least has a workaround with explicit typing.

Reply:
If we can't do a complete job of enforcing consistency maybe we should just remove the constraints completely, as they significantly complicate the expression of the types.

Reply:
@eps1lon as the original author on the react-redux PR, happy to help out if I can. I haven't gotten any reviews in 20 days so it's honestly been a bit hard to iterate/improve on the definition. If there are edge cases it doesn't handle, I can certainly iterate.

I also have my email in my github profile so feel free to reach out that way.

Reply:
FWIW, is the issue you're dealing with related to this by any chance https://github.com/DefinitelyTyped/DefinitelyTyped/issues/28249?

Reply:
That issue is probably causing the fail with the explicitly typed SFC. So if that issue is resolved we might not even have a consistency check for explicitly typed SFCs.